zrml/market-commons/src/migrations.rs

```diff
@@ -14,3 +14,370 @@
+        let mut new_markets_data: Vec<(Vec<u8>, MarketOf<T>)> = Vec::new();
+        for (key, legacy_market) in storage_iter::<LegacyMarketOf<T>>(MARKET_COMMONS, MARKETS) {
+            total_weight = total_weight.saturating_add(T::DbWeight::get().reads(1));
+            let new_market = Market {
+                base_asset: Asset::Ztg,
+                creator: legacy_market.creator,
+                creation: legacy_market.creation,
+                creator_fee: legacy_market.creator_fee,
+                oracle: legacy_market.oracle,
+                metadata: legacy_market.metadata,
+                market_type: legacy_market.market_type,
+                period: legacy_market.period,
+                deadlines: legacy_market.deadlines,
+                scoring_rule: legacy_market.scoring_rule,
+                status: legacy_market.status,
+                report: legacy_market.report,
+                resolved_outcome: legacy_market.resolved_outcome,
+                dispute_mechanism: legacy_market.dispute_mechanism,
+            };
+            new_markets_data.push((key, new_market));
+        }
+        for (key, new_market) in new_markets_data {
+            put_storage_value::<MarketOf<T>>(MARKET_COMMONS, MARKETS, &key, new_market);
+            total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
+        }
+        log::info!("Completed updates of markets");
+        utility::put_storage_version_of_market_commons_pallet(MARKET_COMMONS_NEXT_STORAGE_VERSION);
+        total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
+        total_weight
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<(), &'static str> {
+        use alloc::string::ToString;
+        use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+        use scale_info::prelude::format;
+        let legacy_markets_count_key = "legacy_markets_count_key".to_string();
+        Self::set_temp_storage(0_u32, &legacy_markets_count_key);
+        for (market_id, legacy_market) in storage_iter::<LegacyMarketOf<T>>(MARKET_COMMONS, MARKETS)
+        {
+            Self::set_temp_storage(legacy_market, &format!("{:?}", market_id.as_slice()));
+            let legacy_markets_count: u32 = Self::get_temp_storage(&legacy_markets_count_key)
+                .expect("legacy_markets_count_key storage not found");
+            Self::set_temp_storage(legacy_markets_count + 1_u32, &legacy_markets_count_key);
+        }
+        Ok(())
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade() -> Result<(), &'static str> {
+        use alloc::string::ToString;
+        use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+        use scale_info::prelude::format;
+        let mut markets_count = 0_u32;
+        let legacy_markets_count_key = "legacy_markets_count_key".to_string();
+        for (market_id, updated_market) in storage_iter::<MarketOf<T>>(MARKET_COMMONS, MARKETS) {
```

Comment:
okay, are there any other places in code too where we should use MarketCommonsPalletApi to make sure we are reading correct storage item? like `on_runtime_upgrade()` or `pre_upgrade()` ?

Reply:
Well, in `pre_upgrade` you're writing `LegacyMarket` objects, you don't have any other choice there. In the actualy migration, I think it's fine to use `put_storage_value` even when writing `Market` objects, because we're actually testing that code. My only concern is that tests shouldn't be "self-referential" (not sure if that's even the right word here). I believe one of my recent migrations had that problem.